Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -19,96 +19,366 @@
+	const int32 SpawnAreaHeight = ZoneHeight - MinDistanceBetweenClusters;
 
-	if (SpawnPoints.Num() < ExpectedPlayers) // SpawnPoints can be rounded up if ExpectedPlayers % NumClusters != 0
+	const int32 NumZones = ZoneRows * ZoneCols;
+	const int32 NumBoundaries = ZoneRows * (ZoneCols - 1) + ZoneCols * (ZoneRows - 1);
+
+	const int32 MaxClustersPerZone = FMath::CeilToInt(ZoneClusters / static_cast<float>(NumZones));
+	const int32 MaxClustersPerBoundary = NumBoundaries != 0 ? FMath::CeilToInt(BoundaryClusters / static_cast<float>(NumBoundaries)) : 0;
+
+	for (int32 Row = 0; Row < SpawnAreaRows; ++Row)
 	{
-		UE_LOG(LogBenchmarkGymGameMode, Error, TEXT("Error creating spawnpoints, number of created spawn points (%d) does not equal total players (%d)"), SpawnPoints.Num(), ExpectedPlayers);
+		for (int32 Col = 0; Col < SpawnAreaCols; ++Col)
+		{
+			bool bZoneRow = Row % 2 == 0;
+			bool bZoneCol = Col % 2 == 0;
+			if (!bZoneRow && !bZoneCol)
+			{
+				// Skip boundary intersections
+				continue;
+			}
+
+			bool bIsZone = bZoneRow && bZoneCol;
+			const int32 MaxClusters = bIsZone ? MaxClustersPerZone : MaxClustersPerBoundary;
+			if (MaxClusters == 0)
+			{
+				// Skip areas that need 0 clusters
+				continue;
+			}
+
+			USpawnArea* NewSpawnArea = NewObject<USpawnArea>(this);
+			if (NewSpawnArea != nullptr)
+			{
+				NewSpawnArea->Type = bIsZone ? USpawnArea::Type::Zone : USpawnArea::Type::Boundary;
+
+				const float X = StartX + Col * ZoneWidth / 2;
+				const float Y = StartY + Row * ZoneHeight / 2;
+				NewSpawnArea->WorldPosition = { X, Y, 0.0f };
+
+				NewSpawnArea->Width = bIsZone ? SpawnAreaWidth : bZoneRow ? MinDistanceBetweenClusters : SpawnAreaWidth;
+				NewSpawnArea->Height = bIsZone ? SpawnAreaHeight : bZoneCol ? MinDistanceBetweenClusters : SpawnAreaHeight;
+				NewSpawnArea->MaxClusters = MaxClusters;
+				NewSpawnArea->MaxSpawnPointsPerCluster = MaxSpawnPointsPerCluster;
+				NewSpawnArea->MinDistanceBetweenClusters = MinDistanceBetweenClusters;
+
+				NewSpawnArea->GenerateSpawnClusters();
+				SpawnAreas.Add(NewSpawnArea);
+			}
+		}
 	}
+}
 
-	GenerateTestScenarioLocations();
+void USpawnManager::CreateSpawnPointActors(int32 ZoneClusters, int32 BoundaryClusters, const int32 MaxSpawnPointsPerCluster)
+{
+	for (USpawnArea* SpawnArea : SpawnAreas)
+	{
+		if (SpawnArea->Type == USpawnArea::Type::Boundary && BoundaryClusters > 0)
+		{
+			SpawnPointActors += SpawnArea->CreateSpawnPointActors();
+			BoundaryClusters--;
+		}
+		else if (SpawnArea->Type == USpawnArea::Type::Zone && ZoneClusters > 0)
+		{
+			SpawnPointActors += SpawnArea->CreateSpawnPointActors();
+			ZoneClusters--;
+		}
+	}
+}
 
-	SpawnNPCs(TotalNPCs);
+// --- ABenchmarkGymGameMode ---
+
+ABenchmarkGymGameMode::ABenchmarkGymGameMode()
+	: DistBetweenClusterCenters(40000) // 400 meters, in Unreal units.
+	, PercentageSpawnPointsOnWorkerBoundaries(0.25f)
+	, bHasCreatedSpawnPoints(false)
+	, PlayerDensity(-1)
```

Comment:
What is PlayerDensity -1

Reply:
Added a comment but essentially it is invalid until it is set.